Fast heuristic PCoA (FSVD)

HannesHolste opened the pull request:
Requesting reviews from @antgonza and @wasade, who are briefed on this research project.

Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [x] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.
* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

-> Contains code from our other, internal biocore repository [https://github.com/biocore/mds-approximations](https://github.com/biocore/mds-approximations)


**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

Overview:
We faced a severe PCoA bottleneck, which mds-approximations project sought to understand and overcome. This PR implements FSVD, which was decided to be the candidate for fast, approximate eigendecomposition. FSVD was implemented based on research outlined by Halko et al (2011) in a paper mentioned in the documentation of the pcoa class in this PR.

Comments about PR:
* Added fsvd as an optional method to the pcoa class. But by default, the old eigh is used, which computes "exact" eigenvectors and eigenvalues. No backwards-incompatible changes made.
* Wrote unit tests for newly introduced functions.

HannesHolste commented:
Made a slight improvement, but when running `make test` locally, it fails due to an issue related to the previously merged permdisp #1532 branch... please advise:
```
File "scikit-bio/skbio/stats/distance/_permdisp.py", line 194, in skbio.stats.distance._permdisp.permdisp
Failed example:
    permdisp(dm, df, 'Grouping', permutations=99, test='centroid')
Expected:
    method name               PERMDISP
    test statistic name        F-value
    sample size                      6
    number of groups                 2
    test statistic             3.67082
    p-value                       0.29
    number of permutations          99
    Name: PERMDISP results, dtype: object
Got:
    method name               PERMDISP
    test statistic name        F-value
    sample size                      6
    number of groups                 2
    test statistic             3.67082
    p-value                       0.35
    number of permutations          99
    Name: PERMDISP results, dtype: object


----------------------------------------------------------------------
```

antgonza commented:
Can you push those changes to see if it also fails here? Thanks!

ElDeveloper commented:
@nreeve17, can you have a look at why this is happening?

On (May-01-18|18:12), Hannes Holste wrote:
>Made a slight improvement, but when running `make test` locally, it fails due to an issue related to the previously merged permdisp #1532 branch... please advise:
>```
>File "scikit-bio/skbio/stats/distance/_permdisp.py", line 194, in skbio.stats.distance._permdisp.permdisp
>Failed example:
>    permdisp(dm, df, 'Grouping', permutations=99, test='centroid')
>Expected:
>    method name               PERMDISP
>    test statistic name        F-value
>    sample size                      6
>    number of groups                 2
>    test statistic             3.67082
>    p-value                       0.29
>    number of permutations          99
>    Name: PERMDISP results, dtype: object
>Got:
>    method name               PERMDISP
>    test statistic name        F-value
>    sample size                      6
>    number of groups                 2
>    test statistic             3.67082
>    p-value                       0.35
>    number of permutations          99
>    Name: PERMDISP results, dtype: object
>
>
>----------------------------------------------------------------------
>```
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/biocore/scikit-bio/pull/1569#issuecomment-385835972

wasade commented on skbio/stats/ordination/__init__.py:
```diff
@@ -128,10 +128,13 @@
-from ._utils import (mean_and_std, scale, svd_rank, corr, e_matrix, f_matrix)
+from ._utils import (mean_and_std, scale, svd_rank, corr, e_matrix, f_matrix,
+                     e_matrix_optimized, f_matrix_optimized,
+                     center_distance_matrix_optimized)
```
Would it be more accurate to describe them as `_inplace`?

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -6,33 +6,29 @@
 
 @experimental(as_of="0.4.0")
-def pcoa(distance_matrix):
+def pcoa(distance_matrix, method="eigh", to_dimension=None,
```
@gregcaporaso @ebolyen, I'd like to draw your attention to this PR as it expands the API for `pcoa`. 

The motivation for the expansion is to support a heuristic for computing eigenvectors, and builds off an MDS approximation benchmark underway [here](https://github.com/biocore/mds-approximations/). The reason for this work is both performance and necessity: PCoA using `eigh` cannot handle Qiita-sized analyses. We don't know the exact number of samples where it breaks down (as in, produces `nan`, but I'd guess its in the 60-90k sample range. On 100k samples though, the FSVD approach takes only about 2 hours CPU time which is quite impressive.

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -48,6 +44,23 @@ def pcoa(distance_matrix):
+        eigenvectors and eigenvalues for all dimensions. The alternate
+        method, "fsvd", uses faster heuristic eigendecomposition but loses
+        accuracy. The magnitude of accuracy lost is dependent on dataset.
+    to_dimension : number
```
Should this be `number_of_dimensions`?

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -77,22 +90,39 @@ def pcoa(distance_matrix):
 
-    E_matrix = e_matrix(distance_matrix.data)
+    # Center distance matrix, a requirement for PCoA here
+    distance_matrix = center_distance_matrix_optimized(
```
Assuming this is the updated inplace operation, then it would be imperative to note this in the changelog. If this is inplace, then it silently breaks the API as the existing code did not operate inplace. This can very easily lead to bugs in callers that are using older versions of skbio. Would it be possible to add in a kwarg for using inplace or not?

HannesHolste replied:
Excellent, excellent catch. Thank you.

wasade replied:
Yes, thanks

HannesHolste replied:
Ok, inplace made a kwarg, `False` by default. Also added an additional unit test for `center_distance_matrix` method -- one tests inplace and one non-inplace

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -77,22 +90,39 @@ def pcoa(distance_matrix):
-    # is zero). cogent makes eigenvalues positive by taking the
+    # If no dimension specified, by default will compute all eigenvectors
+    # and eigenvalues
+    if to_dimension is None:
```
should this just be set in the `method == 'eigh'` block?

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +156,165 @@ def pcoa(distance_matrix):
-        samples=pd.DataFrame(coordinates, index=distance_matrix.ids,
+        samples=pd.DataFrame(coordinates, index=distance_matrix_obj.ids,
                              columns=axis_labels),
         proportion_explained=pd.Series(proportion_explained,
```
Has the proportion explained with fsvd been resolved?

HannesHolste replied:
not yet; pls see other comment thread below

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +156,165 @@ def pcoa(distance_matrix):
+    # only expect square matrices.
+    # Take (conjugate) transpose if necessary, because it makes H smaller,
+    # leading to faster computations
+    # if m < n:
```
Can the commented code be removed please?

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +156,165 @@ def pcoa(distance_matrix):
+    #     distance_matrix = distance_matrix.transpose()
+    #     m, n = distance_matrix.shape
+    if m != n:
+        raise ValueError('FSVD.run(...) expects square distance matrix')
```
This error message doesn't seem accurate?

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +156,165 @@ def pcoa(distance_matrix):
+    # Compute the m * ((i+1)l) product matrix
+    Ut = dot(Q, W)
+
+    if m < n:
```
Since `m == n` is this necessary?

wasade commented on skbio/stats/ordination/_utils.py:
```diff
@@ -196,3 +196,62 @@ def f_matrix(E_matrix):
+    Squares and divides by -2 the input elementwise. Eq. 9.20 in
+    Legendre & Legendre 1998.
+    """
+    distance_matrix = np.array(distance_matrix).astype(np.float)
```
Oh, this doesn't actually look like it's inplace... why not just pull `distance_matrix._data` and operate directly on that?

wasade replied:
isn't the type already ensured with a distance matrix?

HannesHolste replied:
1. Agreed; I removed the redundant `np.array()` recast.
2. I also believed type to already have been ensured with a distance matrix. But the unit tests threw an error that was resolved by coercing types to `np.float` explicitly here.

wasade commented on skbio/stats/ordination/tests/data/PCoA_sample_data_3_fsvd_3d_ref:
```diff
@@ -0,0 +1,22 @@
+0.26252020959180306	0.09043197375300857	0.07177687511060699
+
+Proportion explained	3
+0.6180886481996106	0.2129168512318794	0.16899450056851006
```
Are these proportions reasonably concordant with PCoA using `eigh`?

wasade replied:
...it looks like they sum to 1 here which means PC1 variation will likely change substantially if 3 dimensions are requested vs. 10, right?

HannesHolste replied:
fsvd to 3 dimensions:
eigh produces the following proportions on the same dataset, which all sum to 1:
`0.47397595644762963	0.2781864326674703	0.24783761088490003`

wasade replied:
Is this with or without the squaring?

HannesHolste replied:
With trace and after I fixed erroneous squaring of eigvals:

eigh
```
Eigvals	3
0.5123672604605051	0.30071909442702155	0.26791206600414047
```

fsvd
```
Eigvals	3
0.5123672604605054	0.3007190944270222	0.2679120660041405
```

HannesHolste commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -6,33 +6,29 @@
 from ._ordination_results import OrdinationResults
-from ._utils import e_matrix, f_matrix
-
-# - In cogent, after computing eigenvalues/vectors, the imaginary part
```
Yes, I moved that comment further down, so it's next to the code it's referring to.

HannesHolste commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -48,6 +44,23 @@ def pcoa(distance_matrix):
+        all eigenvectors and eigenvalues. If using fast heuristic
+        eigendecomposition through "fsvd", a desired dimension should be
+        specified.
+    normalize_eigenvectors : bool
```
Good point! It's actually not needed; it's just vector normalization, doesn't change results per se. I added a comment to `pcoa` clarifying that the returned eigenvectors are not normalized into unit vectors, and removed that parameter and its associated code.

antgonza commented:
Note that the current failures, as far as I can tell, are due to flake8:
```
flake8 skbio setup.py checklist.py
skbio/stats/ordination/_principal_coordinate_analysis.py:165:1: W293 blank line contains whitespace
make: *** [test] Error 1
```

HannesHolste commented:
@antgonza: thank you, fixed.

@wasade: kindly take a look at my replies to your comments re:
1. inplace discussion: https://github.com/biocore/scikit-bio/pull/1569#discussion_r185649333
2. implementation of trace: https://github.com/biocore/scikit-bio/pull/1569/files/33b4b289271ba687ad2d00d57c357ef2cdcb30ac#diff-3dc43168bed31052eabc6e5b574f8627

antgonza commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +153,170 @@ def pcoa(distance_matrix):
                                        index=axis_labels))
+
+
+def _fsvd(centered_distance_matrix, dimension=3,
```
Should the default be 10? That's normally what Emperor expects.

antgonza commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +153,170 @@ def pcoa(distance_matrix):
+    # Compute the m * ((i+1)l) product matrix
+    Ut = dot(Q, W)
+
+    # V_fsvd = Vt[:, :num_dimensions_out] # unused
```
Can we remove comment? If not my guess is that num_dimensions_out -> dimension, right?

RNAer commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -44,10 +39,26 @@ def pcoa(distance_matrix):
+        By default, uses SciPy's "eigh", which computes exact
+        eigenvectors and eigenvalues for all dimensions. The alternate
+        method, "fsvd", uses faster heuristic eigendecomposition but loses
+        accuracy. The magnitude of accuracy lost is dependent on dataset.
```
the reference for fsvd should be cited in this docstring. I know you mentioned it in `_fsvd()`, but it should be here too so the users can see it in the public API documentation

RNAer commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +153,169 @@ def pcoa(distance_matrix):
 
-    axis_labels = ['PC%d' % i for i in range(1, eigvals.size + 1)]
+    axis_labels = list(
+        ["PC%d" % i for i in range(1, number_of_dimensions + 1)])
```
this is already a list. remove `list`

RNAer commented on skbio/stats/ordination/_utils.py:
```diff
@@ -196,3 +196,81 @@ def f_matrix(E_matrix):
+    Squares and divides by -2 the input element-wise. Eq. 9.20 in
+    Legendre & Legendre 1998.
+
+    Modified from skbio e_matrix function,
```
the original `e_matrix` is deleted? suggest deleting this line from docstring too. it's confusing to people without background info.

RNAer commented on skbio/stats/ordination/_utils.py:
```diff
@@ -196,3 +196,81 @@ def f_matrix(E_matrix):
+    row and column are subtracted, and the mean of the whole
+    matrix is added. Eq. 9.21 in Legendre & Legendre 1998.
+
+    Modified from skbio e_matrix function,
```
here too. and it's f_matrix

RNAer commented on skbio/util/_testing.py:
```diff
@@ -6,14 +6,13 @@
 import numpy.testing as npt
+import os
 import pandas.util.testing as pdt
+import warnings
```
why this change? Not to clutter the imports, we typically put the import of builtin module in top; an empty line; the import of external pkg; an empty line; and then the import of the module in current pkg.

HannesHolste commented:
@RNAer thanks for your feedback, Zech.

@wasade: kindly take a look at my replies to your comments and confirm
1. inplace discussion: https://github.com/biocore/scikit-bio/pull/1569#discussion_r185649333
2. implementation of trace: https://github.com/biocore/scikit-bio/pull/1569#discussion_r185702245

wasade commented:
Build errors look real but unrelated from what I can see. @HannesHolste, can you update `setup.py` and change `'pandas >= 0.19.2',` to `'pandas >= 0.19.2, < 0.23.0',` and see if that cleans this up? 

@ebolyen @thermokarst, can you take a look this specific [`TabularMSA`](https://travis-ci.org/biocore/scikit-bio/jobs/380551571#L1325-L1343) failure with pandas as an example of the problem? I'm guessing this has to do with pandas 0.23 which was released on the 15th, and I'm also guessing we may need to force pandas < 0.23 until this is resolved. Not sure if there is downstream impact with Q2.

ebolyen commented:
@wasade, I think q2 is ok since its environments are fixed, however we're super behind on the general scipy-stack, so we should see what we can do to get back up to date. Looking at the stack-trace there, it seems like we're hitting a branch we never used to hit, something about indexing with a bool-list is breaking. Perhaps one of the properties on the index has changed. TabularMSA tries to handle multi-index correctly but that state is usually pretty ambiguous. I would probably start there for figuring out what's going on.

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,15 +156,169 @@ def pcoa(distance_matrix):
+
+    U_fsvd = Ut[:, :dimension]
+
+    S = St[:dimension] ** 2
```
@ElDeveloper and I discovered that this squaring is incorrect. Is this part of the method description? Based on our testing, this should simply be `S = St[:dimension]`, and the resulting eigenvalues are then directly relatable with minor deviation to the eigenvalues from classic PCoA

wasade commented:
@HannesHolste are you able to resolve the merge conflicts and outstanding comments? It would be great to get this PR completed and merged!

HannesHolste commented:
@wasade Agreed! Carving out time in the evenings & this weekend to tackle them. When is the next skbio release?

wasade commented:
The last one just happened a few days ago. If you can get this in in the
next week, than it's at least possible we can do a release before the next
QIIME2 one

On Thu, Aug 9, 2018 at 3:43 PM Hannes Holste <notifications@github.com>
wrote:

> @wasade <https://github.com/wasade> Agreed! Carving out time in the
> evenings & this weekend to tackle them. When is the next skbio release?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1569#issuecomment-411921184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAc8sm6Kb1sVlXREUT3rnOM0VAaZFcIdks5uPLr6gaJpZM4Tuo2k>
> .
>

HannesHolste commented:
@wasade @ElDeveloper thank you for catching the squaring of the eigenvalues. As far as I can tell, that should not have been there. 

I resolved merge conflicts with master, rewrote the fsvd unit test, made center_distance_matrix optionally inplace (and added new test for that), improved docs, and just tested the results of using the trace method with the squaring issue fixed (https://github.com/biocore/scikit-bio/pull/1569#discussion_r185702245)

Ready for review and merge, if OK.

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +162,174 @@ def pcoa(distance_matrix):
+        # compute the m x l matrices H^{(0)}, ..., H^{(i)}
+        # Note that this is done implicitly in each iteration below.
+        H = dot(centered_distance_matrix, G)
+        # Again, removed transpose: dot(distance_matrix.transpose(), H)
```
Would it be possible to remove the comments about removing code? see also lines 291 295, and maybe others

HannesHolste replied:
All references to removed code in comments deleted, thanks. I only kept the below comment, because it emphasizes that we expect square matrices.

```
    # Note: a (conjugate) transpose is removed for performance, since we
    # only expect square matrices.
    if m != n:
        raise ValueError('FSVD expects square distance matrix')
```

HannesHolste commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +162,170 @@ def pcoa(distance_matrix):
 
 
+def _fsvd(centered_distance_matrix, dimension=10,
+          use_power_method=False, num_levels=1):
```
A use case where an skbio user would configure these seems unlikely / not immediately apparent to me, so I've removed them from the list of function parameters that a user can specify.

HannesHolste commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +162,170 @@ def pcoa(distance_matrix):
+           centered_distance_matrix : np.array
+               Numpy matrix representing the distance matrix for which the
+               eigenvectors and eigenvalues shall be computed
+           dimension : int
```
Great catch, thanks

HannesHolste commented on skbio/stats/ordination/__init__.py:
```diff
@@ -129,10 +129,13 @@
 __all__ = ['ca', 'rda', 'cca', 'pcoa', 'pcoa_biplot', 'OrdinationResults',
            'mean_and_std', 'scale', 'svd_rank', 'corr',
-           'e_matrix', 'f_matrix']
+           'e_matrix', 'f_matrix', 'e_matrix_inplace', 'f_matrix_inplace',
```
Thanks for explicitly pointing this out, that was easy to miss during my refactor.

ElDeveloper replied:
👌

HannesHolste commented on skbio/stats/ordination/_utils.py:
```diff
@@ -196,3 +196,86 @@ def f_matrix(E_matrix):
     matrix_mean = E_matrix.mean()
     return E_matrix - row_means - col_means + matrix_mean
+
+
```
These existed as utility methods before, making them private now might break others' code.

ebolyen replied:
As long as they were never exported from _utils.py there isn't a contract. We would consider a "public function" in a "private package" to be package private (usable within skbio only). That being said, I'm relatively unfamiliar with this section of the codebase, so that "rule" may not really apply here.

HannesHolste replied:
@ebolyen and @ElDeveloper I tried making the f_matrix and e_matrix (and f_matrix_inplace and e_matrix_inplace) methods private, but when I removed the references from skbio.ordination.__init__.py 

```
# I tried removing the f/e matrix method names and import from below
from ._utils import (mean_and_std, scale, svd_rank, corr, e_matrix, f_matrix,
                     center_distance_matrix, _e_matrix_inplace,
                     _f_matrix_inplace)

__all__ = ['ca', 'rda', 'cca', 'pcoa', 'pcoa_biplot', 'OrdinationResults',
           'mean_and_std', 'scale', 'svd_rank', 'corr',
           'e_matrix', 'f_matrix', 'center_distance_matrix',
           '_e_matrix_inplace', '_f_matrix_inplace']
```

HannesHolste replied:
@ElDeveloper ah, thanks for clarifying, you're right, you can import directly from the utils file. I've made those methods private now and removed the references from the `__init__.py`

HannesHolste commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -79,20 +102,33 @@ def pcoa(distance_matrix):
+
+    # If no dimension specified, by default will compute all eigenvectors
+    # and eigenvalues
+    if number_of_dimensions is None:
```
Ah, good catch. Thanks.

HannesHolste commented on skbio/stats/ordination/tests/test_principal_coordinate_analysis.py:
```diff
@@ -53,6 +53,34 @@ def test_simple(self):
+                                        ignore_directionality=True,
+                                        ignore_method_names=True)
+
+    def test_fsvd(self):
```
good idea!

HannesHolste commented on skbio/stats/ordination/tests/test_principal_coordinate_analysis.py:
```diff
@@ -53,6 +53,34 @@ def test_simple(self):
+        assert_ordination_results_equal(results, expected_results,
+                                        ignore_directionality=True,
+                                        ignore_method_names=True)
+
```
Will do, thanks!

HannesHolste commented:
Thanks a ton for your detailed feedback, @ElDeveloper! Changes integrated.

wasade commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -62,37 +82,52 @@ def pcoa(distance_matrix):
+
+        eigvals, eigvecs = eigh(matrix_data)
+        long_method_name = "Principal Coordinate Analysis"
+    elif method == "fsvd":
```
Should we warn users if the number of dimensions is large when computing fsvd? If `number_of_dimensions` is `None` as is default, fsvd will compute a lot of dimensions which may be unexpected for the user user.

HannesHolste replied:
Ok, I added a warning. "Large" is a bit hard to define, as any integer cutoff I add would feel a bit arbitrary, so I just added a general runtime warning of this behavior if the user invokes `fsvd` without specifying `number_of_dimensions`.

wasade replied:
I think the warning is in the wrong place? It looks like this warning will trigger even if the number of dimensions requested is 3?

HannesHolste replied:
Change: the warning will now trigger of dimensions of matrix > 10.

wasade commented on skbio/stats/ordination/_utils.py:
```diff
@@ -196,3 +196,86 @@ def f_matrix(E_matrix):
+
+    Note: If the used distance was euclidean, pairwise distances
+    needn't be computed from the data table Y because F_matrix =
+    Y.dot(Y.T) (if Y has been centred).
```
centred -> centered

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -62,37 +84,60 @@ def pcoa(distance_matrix):
+    Principal Coordinate Analysis is sometimes known as
+    metric multidimensional scaling or classical scaling.
+
+    If the distance is not euclidean (for example if it is a
```
This portion should be left as it was before i.e. under the `.. note::` directive. The only thing that should have been removed from the `.. note::` directive is the sentence "Principal coordinate analysis ....".

HannesHolste replied:
ah, thanks for clarifying

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -44,10 +40,36 @@ def pcoa(distance_matrix):
+        specified `distance_matrix`.
+    inplace : bool, optional
+        If true, centers a distance matrix in-place in a manner that reduces
+        computational and memory cost.
```
This is a comment I had before, can you address it:

> As far as I can tell the inplace operations are done iteratively instead of using a vectorized operation, hence the benefits would only be limited memory not compute. I suggest changing this to "reduces memory consumption".

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -44,10 +40,36 @@ def pcoa(distance_matrix):
+        By default, equal to the number of dimensions of the distance matrix,
+        as default eigendecompsition using SciPy's `eigh` method computes
+        all eigenvectors and eigenvalues. If using fast heuristic
+        eigendecomposition through `fsvd`, a desired dimension should be
```
desired dimension -> desired number of dimensions

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -62,37 +84,60 @@ def pcoa(distance_matrix):
+        # eigh does not natively support specifying number_of_dimensions, i.e.
+        # there are no speed gains unlike in FSVD. Later, we slice off unwanted
+        # dimensions to conform the result of eigh
+        # to the specified number_of_dimensions.
```
Can you remove the line break after eigh? There's still space in line 127 for the text in line 128.

HannesHolste replied:
not quite enough space, but I've reformatted it.

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +171,181 @@ def pcoa(distance_matrix):
 
+    if method == "fsvd":
+        # Since the dimension parameter, hereafter referred to as 'd',
+        # restricts the number of eigenvalues and
```
Same comment as before, can you remove the new line and reindent this comment block. I suggest doing a pass through the rest of the source code looking for this.

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +171,181 @@ def pcoa(distance_matrix):
+
+    Notes
+    -----
+
```
Can you remove this blank line?

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -130,20 +171,181 @@ def pcoa(distance_matrix):
+
+    k = number_of_dimensions + 2
+
+    # Form a real nxl matrix G whose entries are independent,
```
As per my comment before:

> Can you reindent this block of comments to remove the unnecessary line breaks?
Show outdated

HannesHolste replied:
Fixed

ElDeveloper commented on skbio/stats/ordination/_principal_coordinate_analysis.py:
```diff
@@ -62,37 +84,60 @@ def pcoa(distance_matrix):
+    # and eigenvalues
+    if number_of_dimensions == 0:
+        if method == "fsvd" and matrix_data.shape[0] > 10:
+            warn("FSVD: since no value for number_of_dimensions is specified, "
```
Can you add a test that this warning is raised correctly? I believe numpy's testing module has some functionality to do this.

HannesHolste replied:
Great suggestion. I found it, it's in python unittest:

`self.assertWarnsRegex(RuntimeWarning, "expected text")`

HannesHolste commented:
@ElDeveloper OK, addressed formatting and adding a new test! 👍 Travis CI has passed.

ElDeveloper commented:
Thanks @HannesHolste and everyone who provided very useful feedback!

wasade commented:
🎉 That's fantastic!!!
